tests/integrations/java/test_verify_jdk.py

```diff
@@ -0,0 +1,346 @@
+            'java/Contents/Home'
+        ),
+        (
+            "Linux",
+            "https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/"
+            "jdk8u242-b08/OpenJDK8U-jdk_x64_linux_hotspot_8u242b08.tar.gz",
+            'java'
+        ),
+        (
+            "Windows",
+            "https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/"
+            "jdk8u242-b08/OpenJDK8U-jdk_x64_windows_hotspot_8u242b08.zip",
+            'java'
+        ),
+    ]
+)
+def test_successful_jdk_download(test_command, tmp_path, host_os, jdk_url, jhome):
+    "If needed, a JDK can be downloaded."
+    # Mock host OS
+    test_command.host_os = host_os
+
+    # Mock a JAVA_HOME that won't exist
+    # This is only needed to make macOS *not* run /usr/libexec/java_home
+    test_command.os.getenv = mock.MagicMock(return_value='/does/not/exist')
+
+    # Mock the cached download path
+    archive = mock.MagicMock()
+    archive.__str__.return_value = '/path/to/download.zip'
+    test_command.download_url.return_value = archive
+
+    # Create a directory to make it look like Java was downloaded and unpacked.
+    (tmp_path / 'tools' / 'jdk8u242-b08').mkdir(parents=True)
+
+    # Invoke the verify call
+    java_home = verify_jdk(cmd=test_command)
+
+    assert java_home == tmp_path / 'tools' / jhome
+
+    # Download was invoked
+    test_command.download_url.assert_called_with(
+        url=jdk_url,
+        download_path=tmp_path / "tools",
+    )
+    # The archive was unpacked
+    test_command.shutil.unpack_archive.assert_called_with(
+        '/path/to/download.zip',
+        extract_dir=str(tmp_path / "tools")
+    )
+    # The original archive was deleted
+    archive.unlink.assert_called_once_with()
+
+
+def test_jdk_download_failure(test_command, tmp_path):
+    "If an error occurs downloading the JDK, an error is raised"
+    # Mock Linux as the host
+    test_command.host_os = 'Linux'
+
+    # Mock a failure on download
+    test_command.download_url.side_effect = requests_exceptions.ConnectionError
+
+    # Invoking verify_jdk causes a network failure.
+    with pytest.raises(NetworkFailure):
+        verify_jdk(cmd=test_command)
+
+    # That download was attempted
+    test_command.download_url.assert_called_with(
+        url="https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/"
+            "jdk8u242-b08/OpenJDK8U-jdk_x64_linux_hotspot_8u242b08.tar.gz",
+        download_path=tmp_path / "tools",
+    )
+    # No attempt was made to unpack the archive
+    assert test_command.shutil.unpack_archive.call_count == 0
+
+
+def test_invalid_jdk_archive(test_command, tmp_path):
+    "If the JDK download isn't a valid archive, raise an error"
+    # Mock Linux as the host
+    test_command.host_os = 'Linux'
+
+    # Mock the cached download path
+    archive = mock.MagicMock()
+    archive.__str__.return_value = '/path/to/download.zip'
```

Comment:
I find the use of `__str__` confusing here, personally. Can the `return_value` be a `Path` object, or an actual string?

Reply:
As above; not without faking some actual files.